std/meta/index.zig

```diff
@@ -0,0 +1,691 @@
+    debug.assert(codeface == 0);
+    
+    const S = packed struct.
+    {
+        a: u8,
+        b: u8,
+        c: u8,
+        d: u8,
+    };
+    
+    const inst = S.{ .a = 0xBE, .b = 0xEF, .c = 0xDE, .d = 0xA1, };
+    debug.assert(std.mem.eql(u8, bytes(&inst), "\xBE\xEF\xDE\xA1"));
+}
+
+/// Provides the size of a type padded to the nearest byte, instead of however the
+///  compiler actually pads it. Useful for determining the size of a type as it would
+///  likely be stored to disk.
+pub fn byteAlignedSizeOf(comptime T: type) usize
+{
+    comptime
+    {
+        switch(@typeId(T))
+        {   
+            builtin.TypeId.Bool,
+            builtin.TypeId.Int,
+            builtin.TypeId.Float => return @sizeOf(T),
+            builtin.TypeId.Enum => return @sizeOf(@TagType(T)),
+            builtin.TypeId.Array =>
+            {
+                return byteAlignedSizeOf(Child(T)) * T.len;
+            },
+            builtin.TypeId.Struct =>
+            {
+                const info = @typeInfo(T).Struct;
+                
+                if(info.layout == builtin.TypeInfo.ContainerLayout.Packed) return @sizeOf(T);
+                
+                var size = usize(0);
+                for(info.fields) |field|
+                {
+                    size += byteAlignedSizeOf(field.field_type);
+                }
+                return size;
+            },
+            builtin.TypeId.Union =>
+            {
+                const info = @typeInfo(T).Union;
+                var largest = usize(0);
+                for(info.fields) |field|
+                {
+                    const size = byteAlignedSizeOf(field.field_type);
+                    if(size > largest) largest = size;
+                }
+                return largest;
+            },
+            else => return 0,
+        }
+    }
+}
+
+test "std.meta.byteAlignedSizeOf"
+{
+    const UnpackedS = struct.
+    {
+        a: u24,
+        b: u8,
+    };
+
+    debug.assert(byteAlignedSizeOf(UnpackedS) == 4);
+    
+    const UnpackedWithSub = struct.
+    {
+        a: u8,
+        b: u8,
+        unpacked_s: UnpackedS,
+    };
+    
+    debug.assert(byteAlignedSizeOf(UnpackedWithSub) == 4 + 2);
+    
+    const PackedS = packed struct.
+    {
+        b0: bool,
+        b1: bool,
+        b2: bool,
+        b3: bool,
+        b4: bool,
+        b5: bool,
+        b6: bool,
+        b7: bool,
+    };
+    
+    const UnpackedWithPackedSub = struct.
+    {
+        a: u8,
+        b: u8,
+        unpacked_s: UnpackedS,
+        packed_s: PackedS,
+    };
+    
+    debug.assert(byteAlignedSizeOf(UnpackedWithPackedSub) == 4 + 2 + 1);
+}
+
+///Compares two structs or (tagged) unions for equality on a field level
+/// so that padding bytes are not relevant.
+pub fn containerEql(container_a: var, container_b: var) bool
+{
+    const T = UnwrapContainer(@typeOf(container_a));
+    debug.assert(UnwrapContainer(@typeOf(container_b)) == T);
```

Comment:
Again, I don't think we should take pointers. Let the function take values, and let the user deref.